File: 18-4sum/18-4sum.cpp
class Solution {
public:
    vector<vector<int>> twoSum(vector<int>& nums, int si, int target){
        
        vector<vector<int>> ans;
        
        int i=si;
        int j=nums.size()-1;
        
        while(i<j){
            
            if(i!=si && nums[i] == nums[i-1]){
                i++;
                continue;
            }
            
            if(nums[i]+nums[j]==target){
                vector<int> res;
                res.push_back(nums[i]);
                res.push_back(nums[j]);
                ans.push_back(res);
                
                i++;
                j--;
            }
            
            else if(nums[i] + nums[j] > target){
                j--;
            }
            else i++;
        }
        
        return ans;
    }
    
    vector<vector<int>> threeSum(vector<int>& nums, int si, int target){
        
        vector<vector<int>> ans;
        
        for(int i=si; i<nums.size()-2; i++){
            
            if(i!=si && nums[i] == nums[i-1]){
                continue;
            }
            
            vector<vector<int>> res = twoSum(nums, i+1, target-nums[i]);
            
            for(vector<int> v : res){
                v.insert(v.begin(), nums[i]);
                ans.push_back(v);
            }
        }
        
        return ans;
    }
    
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        
        vector<vector<int>> ans;
        
        if(nums.size()==0 || nums.size()==1 || nums.size()==2 || nums.size()==3) return ans;
        
        sort(nums.begin(), nums.end());
        
        for(int i=0; i<nums.size()-3; i++){
            
            if(i!=0 && nums[i] == nums[i-1]){
                continue;
            }
            
            vector<vector<int>> res = threeSum(nums, i+1, target-nums[i]);
            
            for(vector<int> v : res){
                v.insert(v.begin(), nums[i]);
                ans.push_back(v);
            }
        }
        
        return ans;
    }
};

File: 18-4sum/18-4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] arr, int k) {
        
        int n = arr.length;
        
        List<List<Integer>> ans = new ArrayList<>();
        if(n==0 || n==1 || n==2 || n==3) return ans;
    
        Arrays.sort(arr);
        for(int i=0; i<arr.length; i++){
            
            while(i>0 && i<n && arr[i] == arr[i-1]) i++;
            
            for(int j=i+1; j<arr.length; j++){
                
               while(j>i+1 && j<n && arr[j] == arr[j-1]) j++;
                
                int c=j+1;
                int d=arr.length-1;
                
                while(c < d){
                    
                    long sum = (long)arr[i] + (long)arr[j] + (long)arr[c] + (long)arr[d];
                    //System.out.println(sum);
                    if(sum == k){
                        ArrayList<Integer> res = new ArrayList<>();
                        res.add(arr[i]);
                        res.add(arr[j]);
                        res.add(arr[c]);
                        res.add(arr[d]);
                        ans.add(res);
                        c++;
                        
                        while(arr[c] == arr[c-1] && c < d) c++;
                        
                    }
                    
                    else if(sum > k){
                        d--;
                    }
                    
                    else{
                        c++;
                    }
                    
                    
                }
            }
        }
        return ans;
    }
}